pkg/scale/encode.go

```diff
@@ -78,6 +78,12 @@ func (es *encodeState) marshal(in interface{}) (err error) {
 		return
 	}
 
+	vdt, ok := in.(EncodeVaryingDataType)
+	if ok {
+		err = es.encodeVaryingDataType(vdt)
+		return
+	}
+
```

Comment:
for encoding a VDT it only needs to know the variant index, so I would suggest to

```go
type Indexable interface {
    Index() uint
}


vdt, ok := in.(Indexable)
if ok {
	err = es.encodeVaryingDataType(vdt)
	return
}

func (es *encodeState) encodeVaryingDataType(vdt Indexable) (err error) {
	_, err = es.Write([]byte{byte(vdt.Index())})
	if err != nil {
		return
	}
	err = es.marshal(vdt)
	return
}
```

Then you can encode a variant w/o wrapping it around another type, somewhat is done in rust

```rs
let v = RustEnum::Variant
v.encode()
```

in golang

```go
type Variant uint16
func(Variant) Index() uint {return 0}

v := Variant(15)
out, err := scale.Marshal(v)
```

instead of

```go
vdt := MyVaryingDataType{}
err := vdt.SetValue(Variant(15))
if err != nil {
	panic(err)
}

bytes, err := scale.Marshal(vdt)
if err != nil {
	panic(err)
}
```

Reply:
I think that makes sense for the encoding part.  I don't think It can work like that on the decoding side.  We will always need something that is able to provide the codec the types per index.  If there's a use case for just encoding types without wrapping it I think it makes sense to reintroduce the `VaryingDataTypeValue` interface and make the appropriate changes to the codec in a separate feature.